plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+    }
+
+    @Override
+    public void deleteRecursively(URI uri)
+            throws IOException
+    {
+        if (!isDirectory(uri)) {
+            s3.deleteObject(getBucketName(uri), keyFromUri(uri));
+        }
+        else {
+            listDirectories(uri).forEach(dir -> {
+                try {
+                    deleteRecursively(dir);
+                }
+                catch (IOException e) {
+                    e.printStackTrace();
+                }
+            });
+            listFiles(uri).forEach(file -> s3.deleteObject(getBucketName(file), keyFromUri(file)));
+            s3.deleteObject(getBucketName(uri), keyFromUri(uri) + DIRECTORY_SUFFIX);
+        }
+    }
+
+    @Override
+    public Stream<URI> listDirectories(URI dir)
+    {
+        List<String> prefixes = listS3Directory(dir).getCommonPrefixes();
+        return prefixes.stream().map(prefix -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + prefix, dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public Stream<URI> listFiles(URI dir)
+    {
+        List<S3ObjectSummary> objects = listS3Directory(dir).getObjectSummaries();
+        return objects.stream().filter(object -> !object.getKey().endsWith(PATH_SEPARATOR)).map(object -> {
+            try {
+                return new URI(dir.getScheme(), dir.getHost(), PATH_SEPARATOR + object.getKey(), dir.getFragment());
+            }
+            catch (URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        });
+    }
+
+    @Override
+    public long size(URI file)
+            throws IOException
+    {
+        return getS3ObjectMetadata(file).getContentLength();
+    }
+
+    @Override
+    public void close()
+    {
+        uploadExecutor.shutdown();
+        s3.shutdown();
+    }
+
+    ObjectMetadata getS3ObjectMetadata(URI file)
+    {
+        return s3.getObjectMetadata(getBucketName(file), keyFromUri(file));
+    }
+
+    ListObjectsV2Result listS3Directory(URI dir)
+    {
+        String key = keyFromUri(dir);
+        if (!key.isEmpty()) {
+            key += PATH_SEPARATOR;
+        }
+
+        ListObjectsV2Request request = new ListObjectsV2Request()
+                .withBucketName(getBucketName(dir))
+                .withPrefix(key)
+                .withDelimiter(PATH_SEPARATOR);
+        return s3.listObjectsV2(request);
+    }
+
+    private static String keyFromUri(URI uri)
+    {
+        checkArgument(uri.isAbsolute(), "Uri is not absolute: %s", uri);
+        String key = nullToEmpty(uri.getPath());
+        if (key.startsWith(PATH_SEPARATOR)) {
+            key = key.substring(PATH_SEPARATOR.length());
+        }
+        if (key.endsWith(PATH_SEPARATOR)) {
+            key = key.substring(0, key.length() - PATH_SEPARATOR.length());
+        }
```

Comment:
why this trimming? What if we have multiple path separators at the beginning or end?

Reply:
Trimming here is to make sure the semantic equivalent URIs will map to the same s3 key.

> What if we have multiple path separators at the beginning or end?
Actually as long as the naming is consistent, then we are good. However, in implementation we guarantee this doesn't happen.